File: 2017/15/rust/src/main.rs
use std::iter::from_fn;

const MODULUS: u64 = 2147483647;

fn lehmer_rng(m: u64, a: u64, x: u64) -> u64 {
    (a * x) % m
}

fn generator(a: u64, mut x: u64) -> impl Iterator<Item = u64> {
    from_fn(move || {
        x = lehmer_rng(MODULUS, a, x);
        Some(x)
    })
}

fn solve(
    length: usize,
    a: u64,
    pa: impl Fn(&u64) -> bool,
    b: u64,
    pb: impl Fn(&u64) -> bool,
) -> usize {
    generator(16807, a)
        .filter(pa)
        .zip(generator(48271, b).filter(pb))
        .take(length)
        .filter(|(x, y)| (x & 0xffff) == (y & 0xffff))
        .count()
}

fn part1(a: u64, b: u64) -> usize {
    solve(40_000_000, a, |_| true, b, |_| true)
}

fn is_multiple_of(a: u64, b: u64) -> bool {
    (b % a) == 0
}

fn part2(a: u64, b: u64) -> usize {
    solve(
        5_000_000,
        a,
        |&x| is_multiple_of(4, x),
        b,
        |&x| is_multiple_of(8, x),
    )
}

fn main() {
    println!("{}", part1(116, 299));
    println!("{}", part2(116, 299));
}

File: 2017/15/haskell/src/Main.hs
module Main (main) where

import Data.Bits ((.&.))
import Data.Composition ((.:))
import Data.Function (on)

-- INPUT
-- Generator A starts with 116
-- Generator B starts with 299

type Selector = Int -> Bool

lehmerRng :: Int -> Int -> Int -> Int
lehmerRng m a x = (a * x) `mod` m

modulus :: Int
modulus = 2147483647

lsbs :: Int -> Int
lsbs = (.&. 0xffff)

generator :: Int -> Int -> [Int]
generator = drop 1 .: iterate . lehmerRng modulus

solve :: Int -> (Int, Selector) -> (Int, Selector) -> Int
solve len (a, pa) (b, pb) =
  length . filter (== True) . take len $
    zipWith ((==) `on` lsbs)
      (filter pa $ generator 16807 a)
      (filter pb $ generator 48271 b)

part1 :: Int -> Int -> Int
part1 a b = solve 40000000 (a, const True) (b, const True)

isMultipleOf :: Int -> Int -> Bool
isMultipleOf = (== 0) .: flip mod

part2 :: Int -> Int -> Int
part2 a b = solve 5000000 (a, isMultipleOf 4) (b, isMultipleOf 8)

main :: IO ()
main = do
  print $ part1 116 299
  print $ part2 116 299
